contracts/Strategy.sol

```diff
@@ -254,17 +257,18 @@ contract Strategy is BaseStrategy {
                 (uint256 _mToken, ) = MMFarmingPool(mmFarmingPool).userInfo(mmFarmingPoolId, address(this));
                 require(_mToken >= _mvGap, '!insufficientMTokenInMasterChef');
 		
-                MMFarmingPool(mmFarmingPool).withdraw(mmFarmingPoolId, _mvGap);
-                require(IERC20(mmVault).balanceOf(address(this)) >= _mShare, '!mismatchFarmingPoolWithdraw');		
+                MMFarmingPool(mmFarmingPool).withdraw(mmFarmingPoolId, liquidateAll? _mToken : _mvGap);
+                require(IERC20(mmVault).balanceOf(address(this)) >= _mShare, '!mismatchFarmingPoolWithdraw');
             }	
         
-            MMVault(mmVault).withdraw(_mShare);
+            MMVault(mmVault).withdraw(liquidateAll? IERC20(mmVault).balanceOf(address(this)) : _mShare);
             uint256 _after = IERC20(want).balanceOf(address(this));
             require(_after > _before, '!mismatchMushroomsVaultWithdraw');
 
-            _liquidatedAmount = _after.sub(_before);
+            _liquidatedAmount = _after;
             _loss = _amountNeeded > _liquidatedAmount? _amountNeeded.sub(_liquidatedAmount): 0; // due to possible withdrawal fee
-            return (_liquidatedAmount, _loss);		
+            require(_liquidatedAmount + _loss <= _amountNeeded);
+            return (_liquidatedAmount, liquidateAll? 0 : _loss);		
```

Comment:
why is loss 0 if you liquidated all?

Reply:
i think loss should be difference between vault.strategies(address(this)).totalDebt and_liquidatedAmount. You have withdrawn all and we still have debt left that is loss.

Reply:
PR updated to make this `liquidateAll` logic more clear. Thanks for sharing the insight here